[docs-only] ADR-0024 msgpack metadata (#8423)
owncloud/ocis: docs/ocis/adr/0024-msgpack-metadata.md

---
title: "24. Index and store metadata"
date: 2024-02-09T14:57:00+01:00
weight: 24
geekdocRepo: https://github.com/owncloud/ocis
geekdocEditPath: edit/master/docs/ocis/adr
geekdocFilePath: 0024-msgpack-metadata.md
---

* Status: accepted
* Deciders: @butonic, @aduffeck
* Date: [2023-03-13](https://github.com/cs3org/reva/pull/3711/commits/204253eee9dbb8e7fa93a01f3f94a2d28ce40a06)

## Context and Problem Statement

File metadata management is an important aspect for oCIS as a data platform. While using extended attributes to store metadata allows attaching the metadata to the actual file it causes a significant amount of syscalls that outweigh the benefits.

## Decision Drivers <!-- optional -->

> Performance of reading extendet attributes suffers from the syscall overhead when listing and reading all attribudes.

Reviewer: ```suggestion
Performance of reading extended attributes suffers from the syscall overhead when listing and reading all attribudes.
```

> Performance of reading extendet attributes suffers from the syscall overhead when listing and reading all attribudes.

Reviewer 2: ```suggestion
Performance of reading extended attributes suffers from the syscall overhead when listing and reading all attributes.
```

## Considered Options

Going back to the original [ADR-0016 Storage for Files Metadata]({{< ref "0016-files-metadata.md" >}}) we decided to use a dedicated file for metadata storage next to the decomposedfs file representing the node. Several options for the data format were considered:

* Use JSON files to store metadata
* Use INI files to store metadata
* Use msgpack files to store metadata
* Use protobuf messages to store metadata

## Decision Outcome

> Chosen option: "[msgpack files](#msgpack-files)", because we want to stay with a self describing binary format. This is a performance tradeoff that is faster and more efficient than text based formats and more flexible but less efficient then protobuf.

Reviewer 2: ```suggestion
Chosen option: "[msgpack files](#msgpack-files)", because we want to stay with a self describing binary format. This is a performance tradeoff that is faster and more efficient than text based formats and more flexible but less efficient than protobuf.
```
